src/ddsrt/src/sync/freertos/sync.c

```diff
@@ -0,0 +1,505 @@
+  if (len == 0) {
+    ddsrt_free(cond->tasks.buf);
+    (void)memset(&cond->tasks, 0, sizeof(cond->tasks));
+  } else {
+    if (len > cond->tasks.len) {
+      off = cond->tasks.len;
+    } else {
+      off = len;
+    }
+
+    /* Resize buffer if required number of slots is more than available or
+       amount of space freed up is worth the effort. */
+    if (len > cond->tasks.len || len < (cond->tasks.len - 1)) {
+      buf = ddsrt_realloc(cond->tasks.buf, len * sz);
+      if (buf == NULL) {
+        rc = DDS_RETCODE_OUT_OF_RESOURCES;
+      } else {
+        cond->tasks.buf = buf;
+        cond->tasks.len = len;
+      }
+    }
+
+    /* Initialize memory that should not be used anymore/yet to zero. */
+    if (off < cond->tasks.len) {
+      (void)memset(cond->tasks.buf + off, 0, (cond->tasks.len - off) * sz);
+    }
+  }
+
+  return rc;
+}
+
+static dds_retcode_t
+cond_push(ddsrt_cond_t *cond, TaskHandle_t task)
+{
+  size_t cnt;
+
+  assert(cond != NULL);
+  assert(task != NULL);
+
+  /* Ensure the task is not listed. */
+  for (cnt = 0; cnt < cond->tasks.cnt && cond->tasks.buf[cnt] != task; cnt++) {
+    /* Do nothing. */
+  }
+
+  assert(cnt <= cond->tasks.cnt);
+
+  if (cnt == cond->tasks.len && cond_resize(cond, cnt + 1) != DDS_RETCODE_OK) {
+    return DDS_RETCODE_OUT_OF_RESOURCES;
+  }
+
+  cond->tasks.buf[cond->tasks.cnt] = task;
+  cond->tasks.cnt++;
+
+  return DDS_RETCODE_OK;
+}
+
+static void
+cond_pop(ddsrt_cond_t *cond, TaskHandle_t task)
+{
+  size_t cnt;
+
+  assert(cond != NULL);
+  assert(task != NULL);
+
+  /* Locate task index. */
+  for (cnt = 0; cnt < cond->tasks.cnt && cond->tasks.buf[cnt] != task; cnt++) {
+    /* Do nothing. */
+  }
+
+  if (cnt < cond->tasks.cnt) {
+    cond->tasks.cnt--;
+
+    if (cnt < cond->tasks.cnt) {
+      /* Ensure list of tasks is neatly packed. */
+      memmove((cond->tasks.buf +  cnt),
+              (cond->tasks.buf + (cnt + 1)),
+              (cond->tasks.cnt -  cnt) * sizeof(*cond->tasks.buf));
+    }
+
+    (void)cond_resize(cond, cond->tasks.cnt);
+  }
+}
+
+static dds_retcode_t
+cond_timedwait(
+  ddsrt_cond_t *cond,
+  ddsrt_mutex_t *mutex,
+  dds_duration_t reltime)
+{
+  dds_retcode_t rc = DDS_RETCODE_TIMEOUT;
+  dds_time_t abstime;
+  TaskHandle_t task;
+  TickType_t ticks = 0;
+
+  assert(cond != NULL);
+  assert(mutex != NULL);
+
+  abstime = ddsrt_time_add_duration(dds_time(), reltime);
+  ticks = ddsrt_duration_to_ticks_ceil(reltime);
+
+  ddsrt_mutex_lock(&cond->mutex);
+  ddsrt_mutex_unlock(mutex);
+
+  task = xTaskGetCurrentTaskHandle();
+  /* Register current task with condition. */
+  if ((rc = cond_push(cond, task)) == DDS_RETCODE_OK) {
+    /* Discard pending notifications. */
+    (void)ulTaskNotifyTake(1, 0);
+  }
+
+  ddsrt_mutex_unlock(&cond->mutex);
+
+  if (rc == DDS_RETCODE_OK) {
+    /* Wait for notification. */
+    switch (ulTaskNotifyTake(1, ticks)) {
+      case 0:
+        ddsrt_mutex_lock(&cond->mutex);
```

Comment:
Perhaps it would be wise to note here that by the time you call ``cond_pop`` it may already have been removed